src/category_theory/groupoid/subgroupoid.lean

```diff
@@ -0,0 +1,400 @@
+  assuming `v` is contained in the `subgroupoid` (meaning, by definition, that the arrow `𝟙 v`
+  is contained in the subgroupoid).
+
+## Implementation details
+
+The structure of this file is copied from/inspired by `group_theory.subgroup.basic`
+and `combinatorics.simple_graph.subgraph`.
+
+## TODO
+
+* Equivalent inductive characterization of generated (normal) subgroupoids.
+* Characterization of normal subgroupoids as kernels.
+
+## Tags
+
+subgroupoid
+
+-/
+
+open set classical function
+local attribute [instance] prop_decidable
+
+namespace category_theory
+
+namespace groupoid
+
+universes u v
+
+variables {C : Type u} [groupoid C]
+
+
+/--
+A sugroupoid of `C` consists of a choice of arrows for each pair of vertices, closed
+under composition and inverses
+-/
+@[ext] structure subgroupoid (C : Type u) [groupoid C] :=
+  (arrws : ∀ (c d : C), set (c ⟶ d))
+  (inv' : ∀ {c d} {p : c ⟶ d} (hp : p ∈ arrws c d),
+            inv p ∈ arrws d c)
+  (mul' : ∀ {c d e} {p} (hp : p ∈ arrws c d) {q} (hq : q ∈ arrws d e),
+            p ≫ q ∈ arrws c e)
+
+namespace subgroupoid
+
+variable (S : subgroupoid C)
+
+/-- The vertices of `C` on which `S` has non-trivial isotropy -/
+def carrier : set C := {c : C | (S.arrws c c).nonempty }
+
+lemma id_mem_of_nonempty_isotropy (c : C) :
+  c ∈ carrier S → 𝟙 c ∈ S.arrws c c :=
+begin
+  rintro ⟨γ,hγ⟩,
+  have : 𝟙 c = γ * (inv γ), by simp only [vertex_group_mul, comp_inv],
+  rw this,
+  apply S.mul' hγ (S.inv' hγ),
+end
+
+/-- A subgroupoid seen as a quiver on vertex set `C` -/
+def as_wide_quiver : quiver C := ⟨λ c d, subtype $ S.arrws c d⟩
+
+/-- Type synonim for the coercion of a subgroupoid as a groupoid -/
+def coe (S : subgroupoid C) := subtype S.carrier
+
+instance [h : S.carrier.nonempty] : nonempty S.coe := ⟨⟨h.some,h.some_spec⟩⟩
+
+/-- The coercion of a subgroupoid as a groupoid -/
+instance coe_groupoid : groupoid S.coe :=
+{ to_category :=
+  { to_category_struct :=
+    { to_quiver :=
+      { hom := λ a b, S.arrws a.val b.val },
+      id := λ a, ⟨𝟙 a.val, id_mem_of_nonempty_isotropy S a.val a.prop⟩,
+      comp := λ a b c p q, ⟨p.val ≫ q.val, S.mul' p.prop q.prop⟩, },
+    id_comp' := λ a b ⟨p,hp⟩, by simp only [category.id_comp],
+    comp_id' := λ a b ⟨p,hp⟩, by simp only [category.comp_id],
+    assoc' := λ a b c d ⟨p,hp⟩ ⟨q,hq⟩ ⟨r,hr⟩, by simp only [category.assoc] },
+  inv := λ a b p, ⟨inv p.val, S.inv' p.prop⟩,
+  inv_comp' := λ a b ⟨p,hp⟩, by simp only [inv_comp],
+  comp_inv' := λ a b ⟨p,hp⟩, by simp only [comp_inv] }
+
+/-- The subgroup of the vertex group at `c` given by the subgroupoid -/
+def vertex_subgroup {c : C} (hc : c ∈ S.carrier) : subgroup (c ⟶ c) :=
+⟨ S.arrws c c,
+  λ f g hf hg, S.mul' hf hg,
+  by {apply id_mem_of_nonempty_isotropy, use hc,},
+  λ f hf, S.inv' hf⟩
+
+/-- `S` is a subgroupoid of `T` if it is contained in it -/
+def is_subgroupoid (S T : subgroupoid C) : Prop :=
+  ∀ {c d}, S.arrws c d ⊆ T.arrws c d
```

Comment:
In the case of subgroups, the (≤) relation comes from [set_like.partial_order](https://leanprover-community.github.io/mathlib_docs/data/set_like/basic.html#set_like.partial_order). I think you can still make a `set_like` instance using the sigma type over `c` and `d` here (you may or may not want to use [set.sigma](https://leanprover-community.github.io/mathlib_docs/data/set/sigma.html#set.sigma)).

Reply:
I think you automatically get (≤) and everything down to the partial order instance. Instead of `intros c d p` you would do `rintro ⟨c, d, p⟩` (or `λ ⟨c, d, p⟩`) which is almost as convenient as it's now. You can't write `p ∈ S` but should be able to write ` ⟨_, _, p⟩ ∈ S`, and you may `def mk {c d} (p : c ⟶ d) : Σ c d, c ⟶ d := ⟨c, d, p⟩` (or declare a notation) so that you can write `mk p ∈ S`. (A notation is better since `mk` isn't a good name, and `sigma_hom_mk` is too long ...)